Precompiled header with g++

#### Description:

We are currently using precompiled model headers (model_header.hpp.gch) only with clang. See lines 20-23 here: https://github.com/stan-dev/cmdstan/blob/develop/make/program#L20

We could do the same with `g++`  which would speedup `g++` compilation of a model to match with clang - see https://github.com/stan-dev/cmdstan/issues/863#issuecomment-622392026.

I tried this out on Linux and Windows. Here is what I found out (I used the example bernoulli model):

- Ubuntu g++ 7.2.0: compile time goes from 12.5s to 6s
- Windows with Rtools 4.0 (g++ 8): compile time goes from 22s to 13s
- Windows with Rtools 3.5 (g++ 4.9.3): does not work

The precompiled header with g++ does take up 500-700MB (the clang++ one takes 170MB on Linux). There is also a warning due to an Eigen issue that needs to be silenced with` -Wno-ignored-attributes` (see comment below).

Given the size of the precompiled header, the warnings that need to be turned off, the fact that it does not work on all of our supported systems and that we can't detect which compiler we use with makefiles, we have the following options:

- A: use it by default on Linux and Mac (do people use g++ on a Mac at all?), and hide it behind a makefile flag (`PRECOMPILED_HEADERS=true` for example) so users on Windows with RTools 4.0 can use it.
- B: hide it behind a makefile flag for g++ on all OSes
- C: not bother with this, close issue

#### Current Version:
v2.23.0

I almost forgot, there is also a warning with g++:
```
In file included from stan/lib/stan_math/lib/eigen_3.3.7/Eigen/Core:434:0,
                 from stan/lib/stan_math/lib/eigen_3.3.7/Eigen/Dense:1,
                 from stan/lib/stan_math/stan/math/prim/fun/Eigen.hpp:22,
                 from stan/lib/stan_math/stan/math/rev.hpp:4,
                 from stan/lib/stan_math/stan/math.hpp:19,
                 from stan/src/stan/model/model_header.hpp:4:
stan/lib/stan_math/lib/eigen_3.3.7/Eigen/src/Core/DenseCoeffsBase.h: In instantiation of ‘class Eigen::DenseCoeffsBase<Eigen::CwiseUnaryOp<Eigen::internal::scalar_abs2_op<double>, const Eigen::Block<Eigen::Block<Eigen::Matrix<double, -1, -1>, -1, 1, true>, -1, 1, false> >, 0>’:
stan/lib/stan_math/lib/eigen_3.3.7/Eigen/src/Core/DenseBase.h:41:34:   required from ‘class Eigen::DenseBase<Eigen::CwiseUnaryOp<Eigen::internal::scalar_abs2_op<double>, const Eigen::Block<Eigen::Block<Eigen::Matrix<double, -1, -1>, -1, 1, true>, -1, 1, false> > >’
stan/lib/stan_math/lib/eigen_3.3.7/Eigen/src/Core/MatrixBase.h:48:34:   required from ‘class Eigen::MatrixBase<Eigen::CwiseUnaryOp<Eigen::internal::scalar_abs2_op<double>, const Eigen::Block<Eigen::Block<Eigen::Matrix<double, -1, -1>, -1, 1, true>, -1, 1, false> > >’
stan/lib/stan_math/lib/eigen_3.3.7/Eigen/src/Core/CwiseUnaryOp.h:94:7:   required from ‘class Eigen::CwiseUnaryOpImpl<Eigen::internal::scalar_abs2_op<double>, const Eigen::Block<Eigen::Block<Eigen::Matrix<double, -1, -1>, -1, 1, true>, -1, 1, false>, Eigen::Dense>’
stan/lib/stan_math/lib/eigen_3.3.7/Eigen/src/Core/CwiseUnaryOp.h:55:7:   required from ‘class Eigen::CwiseUnaryOp<Eigen::internal::scalar_abs2_op<double>, const Eigen::Block<Eigen::Block<Eigen::Matrix<double, -1, -1>, -1, 1, true>, -1, 1, false> >’
stan/lib/stan_math/lib/eigen_3.3.7/Eigen/src/Core/Dot.h:98:43:   required from ‘typename Eigen::NumTraits<typename Eigen::internal::traits<T>::Scalar>::Real Eigen::MatrixBase<Derived>::squaredNorm() const [with Derived = Eigen::Block<Eigen::Block<Eigen::Matrix<double, -1, -1>, -1, 1, true>, -1, 1, false>; typename Eigen::NumTraits<typename Eigen::internal::traits<T>::Scalar>::Real = double]’
stan/lib/stan_math/stan/math/prim/fun/multiply_lower_tri_self_transpose.hpp:34:47:   required from here
stan/lib/stan_math/lib/eigen_3.3.7/Eigen/src/Core/DenseCoeffsBase.h:55:30: warning: ignoring attributes on template argument ‘Eigen::internal::packet_traits<double>::type {aka __vector(2) double}’ [-Wignored-attributes]
                      >::type PacketReturnType;
                              ^~~~~~~~~~~~~~~~
stan/lib/stan_math/lib/eigen_3.3.7/Eigen/src/Core/DenseCoeffsBase.h: In instantiation of ‘class Eigen::DenseCoeffsBase<Eigen::Transpose<const Eigen::Transpose<const Eigen::Product<Eigen::Matrix<double, -1, -1>, Eigen::TriangularView<const Eigen::Transpose<const Eigen::Matrix<double, -1, -1> >, 2>, 0> > >, 0>’:
stan/lib/stan_math/lib/eigen_3.3.7/Eigen/src/Core/DenseBase.h:41:34:   required from ‘class Eigen::DenseBase<Eigen::Transpose<const Eigen::Transpose<const Eigen::Product<Eigen::Matrix<double, -1, -1>, Eigen::TriangularView<const Eigen::Transpose<const Eigen::Matrix<double, -1, -1> >, 2>, 0> > > >’
stan/lib/stan_math/lib/eigen_3.3.7/Eigen/src/Core/MatrixBase.h:48:34:   required from ‘class Eigen::MatrixBase<Eigen::Transpose<const Eigen::Transpose<const Eigen::Product<Eigen::Matrix<double, -1, -1>, Eigen::TriangularView<const Eigen::Transpose<const Eigen::Matrix<double, -1, -1> >, 2>, 0> > > >’
stan/lib/stan_math/lib/eigen_3.3.7/Eigen/src/Core/Transpose.h:115:37:   required from ‘class Eigen::TransposeImpl<const Eigen::Transpose<const Eigen::Product<Eigen::Matrix<double, -1, -1>, Eigen::TriangularView<const Eigen::Transpose<const Eigen::Matrix<double, -1, -1> >, 2>, 0> >, Eigen::Dense>’
stan/lib/stan_math/lib/eigen_3.3.7/Eigen/src/Core/Transpose.h:52:37:   required from ‘class Eigen::Transpose<const Eigen::Transpose<const Eigen::Product<Eigen::Matrix<double, -1, -1>, Eigen::TriangularView<const Eigen::Transpose<const Eigen::Matrix<double, -1, -1> >, 2>, 0> > >’
stan/lib/stan_math/stan/math/prim/fun/tcrossprod.hpp:27:14:   required from ‘Eigen::Matrix<typename stan::value_type<T>::type, T:: RowsAtCompileTime, T:: RowsAtCompileTime> stan::math::tcrossprod(const T&) [with T = Eigen::Transpose<const Eigen::Product<Eigen::Matrix<double, -1, -1>, Eigen::TriangularView<const Eigen::Transpose<const Eigen::Matrix<double, -1, -1> >, 2>, 0> >; stan::require_eigen_vt<std::is_arithmetic, T>* <anonymous> = 0; typename stan::value_type<T>::type = double]’
stan/lib/stan_math/stan/math/prim/fun/crossprod.hpp:21:20:   required from ‘auto stan::math::crossprod(const EigMat&) [with EigMat = Eigen::Product<Eigen::Matrix<double, -1, -1>, Eigen::TriangularView<const Eigen::Transpose<const Eigen::Matrix<double, -1, -1> >, 2>, 0>; stan::require_eigen_t<T>* <anonymous> = 0]’
stan/lib/stan_math/stan/math/prim/prob/wishart_rng.hpp:30:41:   required from here
stan/lib/stan_math/lib/eigen_3.3.7/Eigen/src/Core/DenseCoeffsBase.h:55:30: warning: ignoring attributes on template argument ‘Eigen::internal::packet_traits<double>::type {aka __vector(2) double}’ [-Wignored-attributes]
stan/lib/stan_math/lib/eigen_3.3.7/Eigen/src/Core/DenseCoeffsBase.h: In instantiation of ‘class Eigen::DenseCoeffsBase<Eigen::Product<Eigen::Transpose<const Eigen::Product<Eigen::Matrix<double, -1, -1>, Eigen::TriangularView<const Eigen::Transpose<const Eigen::Matrix<double, -1, -1> >, 2>, 0> >, Eigen::Transpose<const Eigen::Transpose<const Eigen::Product<Eigen::Matrix<double, -1, -1>, Eigen::TriangularView<const Eigen::Transpose<const Eigen::Matrix<double, -1, -1> >, 2>, 0> > >, 0>, 0>’:
stan/lib/stan_math/lib/eigen_3.3.7/Eigen/src/Core/DenseBase.h:41:34:   required from ‘class Eigen::DenseBase<Eigen::Product<Eigen::Transpose<const Eigen::Product<Eigen::Matrix<double, -1, -1>, Eigen::TriangularView<const Eigen::Transpose<const Eigen::Matrix<double, -1, -1> >, 2>, 0> >, Eigen::Transpose<const Eigen::Transpose<const Eigen::Product<Eigen::Matrix<double, -1, -1>, Eigen::TriangularView<const Eigen::Transpose<const Eigen::Matrix<double, -1, -1> >, 2>, 0> > >, 0> >’
stan/lib/stan_math/lib/eigen_3.3.7/Eigen/src/Core/MatrixBase.h:48:34:   required from ‘class Eigen::MatrixBase<Eigen::Product<Eigen::Transpose<const Eigen::Product<Eigen::Matrix<double, -1, -1>, Eigen::TriangularView<const Eigen::Transpose<const Eigen::Matrix<double, -1, -1> >, 2>, 0> >, Eigen::Transpose<const Eigen::Transpose<const Eigen::Product<Eigen::Matrix<double, -1, -1>, Eigen::TriangularView<const Eigen::Transpose<const Eigen::Matrix<double, -1, -1> >, 2>, 0> > >, 0> >’
stan/lib/stan_math/lib/eigen_3.3.7/Eigen/src/Core/Product.h:115:7:   required from ‘class Eigen::internal::dense_product_base<Eigen::Transpose<const Eigen::Product<Eigen::Matrix<double, -1, -1>, Eigen::TriangularView<const Eigen::Transpose<const Eigen::Matrix<double, -1, -1> >, 2>, 0> >, Eigen::Transpose<const Eigen::Transpose<const Eigen::Product<Eigen::Matrix<double, -1, -1>, Eigen::TriangularView<const Eigen::Transpose<const Eigen::Matrix<double, -1, -1> >, 2>, 0> > >, 0, 8>’
stan/lib/stan_math/lib/eigen_3.3.7/Eigen/src/Core/Product.h:147:7:   required from ‘class Eigen::ProductImpl<Eigen::Transpose<const Eigen::Product<Eigen::Matrix<double, -1, -1>, Eigen::TriangularView<const Eigen::Transpose<const Eigen::Matrix<double, -1, -1> >, 2>, 0> >, Eigen::Transpose<const Eigen::Transpose<const Eigen::Product<Eigen::Matrix<double, -1, -1>, Eigen::TriangularView<const Eigen::Transpose<const Eigen::Matrix<double, -1, -1> >, 2>, 0> > >, 0, Eigen::Dense>’
stan/lib/stan_math/lib/eigen_3.3.7/Eigen/src/Core/Product.h:71:7:   required from ‘class Eigen::Product<Eigen::Transpose<const Eigen::Product<Eigen::Matrix<double, -1, -1>, Eigen::TriangularView<const Eigen::Transpose<const Eigen::Matrix<double, -1, -1> >, 2>, 0> >, Eigen::Transpose<const Eigen::Transpose<const Eigen::Product<Eigen::Matrix<double, -1, -1>, Eigen::TriangularView<const Eigen::Transpose<const Eigen::Matrix<double, -1, -1> >, 2>, 0> > >, 0>’
stan/lib/stan_math/stan/math/prim/fun/tcrossprod.hpp:27:14:   required from ‘Eigen::Matrix<typename stan::value_type<T>::type, T:: RowsAtCompileTime, T:: RowsAtCompileTime> stan::math::tcrossprod(const T&) [with T = Eigen::Transpose<const Eigen::Product<Eigen::Matrix<double, -1, -1>, Eigen::TriangularView<const Eigen::Transpose<const Eigen::Matrix<double, -1, -1> >, 2>, 0> >; stan::require_eigen_vt<std::is_arithmetic, T>* <anonymous> = 0; typename stan::value_type<T>::type = double]’
stan/lib/stan_math/stan/math/prim/fun/crossprod.hpp:21:20:   required from ‘auto stan::math::crossprod(const EigMat&) [with EigMat = Eigen::Product<Eigen::Matrix<double, -1, -1>, Eigen::TriangularView<const Eigen::Transpose<const Eigen::Matrix<double, -1, -1> >, 2>, 0>; stan::require_eigen_t<T>* <anonymous> = 0]’
stan/lib/stan_math/stan/math/prim/prob/wishart_rng.hpp:30:41:   required from here
stan/lib/stan_math/lib/eigen_3.3.7/Eigen/src/Core/DenseCoeffsBase.h:55:30: warning: ignoring attributes on template argument ‘Eigen::internal::packet_traits<double>::type {aka __vector(2) double}’ [-Wignored-attributes]
```
Its an Eigen issue: https://gitlab.com/libeigen/eigen/-/issues/1221 
They propose setting  `-Wno-ignored-attributes` to compile flags. Then it goes away.
It has no effect on performance.

--- a/make/program
+++ b/make/program
@@ -52,12 +52,11 @@ endif
 
 ifeq ($(PRECOMPILED_HEADERS),true)
 PRECOMPILED_MODEL_HEADER=$(STAN)src/stan/model/model_header.hpp.gch
-CXX
-else
-PRECOMPILED_MODEL_HEADER=
 ifeq ($(CXX_TYPE),gcc)
 CXXFLAGS_PROGRAM+= -Wno-ignored-attributes
 endif
+else
+PRECOMPILED_MODEL_HEADER=
 endif
 
 %.d: %.hpp

--- a/make/program
+++ b/make/program
@@ -17,11 +17,6 @@ $(STAN)src/stan/model/model_header.hpp.gch : $(STAN)src/stan/model/model_header.
 	@echo '--- Compiling pre-compiled header ---	'
 	$(COMPILE.cpp) $< $(OUTPUT_OPTION)
 
-ifeq ($(CXX_TYPE),clang)
-CXXFLAGS_PROGRAM += -include-pch $(STAN)src/stan/model/model_header.hpp.gch
-$(STAN_TARGETS) examples/bernoulli/bernoulli$(EXE) $(patsubst %.stan,%$(EXE),$(wildcard src/test/test-models/*.stan)) : %$(EXE) : $(STAN)src/stan/model/model_header.hpp.gch
-endif
-
 ifneq ($(findstring allow_undefined,$(STANCFLAGS)),)
 $(STAN_TARGETS) examples/bernoulli/bernoulli$(EXE) $(patsubst %.stan,%$(EXE),$(wildcard src/test/test-models/*.stan)) : CXXFLAGS_PROGRAM += -include $(USER_HEADER)
 endif
@@ -48,7 +43,7 @@ endif
 %.d: %.hpp
 
 .PRECIOUS: %.hpp
-%$(EXE) : %.hpp $(CMDSTAN_MAIN_O) $(LIBSUNDIALS) $(MPI_TARGETS) $(TBB_TARGETS)
+%$(EXE) : %.hpp $(CMDSTAN_MAIN_O) $(LIBSUNDIALS) $(MPI_TARGETS) $(TBB_TARGETS) $(STAN)src/stan/model/model_header.hpp.gch
 	@echo ''
 	@echo '--- Compiling, linking C++ code ---'
 	$(COMPILE.cpp) $(CXXFLAGS_PROGRAM) -x c++ -o $(subst  \,/,$*).o $(subst \,/,$<)

--- a/make/program
+++ b/make/program
@@ -52,8 +52,12 @@ endif
 
 ifeq ($(PRECOMPILED_HEADERS),true)
 PRECOMPILED_MODEL_HEADER=$(STAN)src/stan/model/model_header.hpp.gch
+CXX
 else
 PRECOMPILED_MODEL_HEADER=
+ifeq ($(CXX_TYPE),gcc)
+CXXFLAGS_PROGRAM+= -Wno-ignored-attributes
+endif
 endif
 
 %.d: %.hpp

--- a/make/program
+++ b/make/program
@@ -44,10 +44,22 @@ endif
 	@echo '--- Translating Stan model to C++ code ---'
 	$(WINE) $(BINSTANC) $(STANCFLAGS) --o=$(subst  \,/,$@) $(subst  \,/,$<)
 
+ifeq ($(OS),Windows_NT)
+PRECOMPILED_HEADERS ?= false
+else
+PRECOMPILED_HEADERS ?= true
+endif
+
+ifeq ($(PRECOMPILED_HEADERS),true)
+PRECOMPILED_MODEL_HEADER=$(STAN)src/stan/model/model_header.hpp.gch
+else
+PRECOMPILED_MODEL_HEADER=
+endif
+
 %.d: %.hpp
 
 .PRECIOUS: %.hpp
-%$(EXE) : %.hpp $(CMDSTAN_MAIN_O) $(LIBSUNDIALS) $(MPI_TARGETS) $(TBB_TARGETS) $(STAN)src/stan/model/model_header.hpp.gch
+%$(EXE) : %.hpp $(CMDSTAN_MAIN_O) $(LIBSUNDIALS) $(MPI_TARGETS) $(TBB_TARGETS) $(PRECOMPILED_MODEL_HEADER)
 	@echo ''
 	@echo '--- Compiling, linking C++ code ---'
 	$(COMPILE.cpp) $(CXXFLAGS_PROGRAM) -x c++ -o $(subst  \,/,$*).o $(subst \,/,$<)

--- a/make/program
+++ b/make/program
@@ -17,6 +17,10 @@ $(STAN)src/stan/model/model_header.hpp.gch : $(STAN)src/stan/model/model_header.
 	@echo '--- Compiling pre-compiled header ---	'
 	$(COMPILE.cpp) $< $(OUTPUT_OPTION)
 
+ifeq ($(CXX_TYPE),clang)
+CXXFLAGS_PROGRAM += -include-pch $(STAN)src/stan/model/model_header.hpp.gch
+endif
+
 ifneq ($(findstring allow_undefined,$(STANCFLAGS)),)
 $(STAN_TARGETS) examples/bernoulli/bernoulli$(EXE) $(patsubst %.stan,%$(EXE),$(wildcard src/test/test-models/*.stan)) : CXXFLAGS_PROGRAM += -include $(USER_HEADER)
 endif

